source/1_engine/rex_engine/src/cmdline.cpp

```diff
@@ -0,0 +1,178 @@
+
+        // the module we're launching is always the first argument on the commandline
+        // we skip this
+        count_t space_pos = cmdLine.find_first_of(' ', start_pos);
+        start_pos         = space_pos + 1;
+
+        // get the new space pos
+        space_pos = cmdLine.find_first_of(' ', start_pos);
+
+        const rsl::string_view arg_prefix = "-"; // all arguments should start with a '-'
+        while(start_pos != -1 && space_pos != -1)
+        {
+          const count_t length            = space_pos - start_pos - arg_prefix.size();
+          const rsl::string_view argument = cmdLine.substr(start_pos + arg_prefix.size(), length);
+          add_argument(argument);
+          start_pos = cmdLine.find_first_not_of(' ', space_pos); // skip all additional spaces
+          space_pos = cmdLine.find_first_of(' ', start_pos);
+        }
+
+        if(start_pos != -1)
+        {
+          add_argument(cmdLine.substr(start_pos + arg_prefix.size())); // + 1 to ignore '-'
+        }
+      }
+
+      void add_argument(rsl::string_view arg)
+      {
+        const count_t equal_pos = arg.find('=');
+        rsl::string_view key    = "";
+        rsl::string_view value  = "";
+
+        // if the argument is of type -MyArg=Something
+        if(equal_pos != -1)
+        {
+          key   = arg.substr(0, equal_pos);
+          value = arg.substr(equal_pos + 1);
+        }
+        // if the argument is of type -EnableSomething
+        else
+        {
+          key   = arg;
+          value = "1"; // this is so we can easily convert to bool/int
+        }
+
+        auto cmd_it = rsl::find_if(g_command_line_args.cbegin(), g_command_line_args.cend(), [key](const CommandLineArgument& cmdArg) { return key == cmdArg.name; });
+
+        if(cmd_it == g_command_line_args.cend())
+        {
+          REX_WARN(LogEngine, "Command {} passed in but it's not recognised as a valid command so will be ignored", arg);
+          return;
+        }
+
+        m_arguments.push_back({key, value});
+      }
+
+      bool verify_args(const CommandLineArgument* args, count_t argCount) // NOLINT(readability-convert-member-functions-to-static)
+      {
+        for(count_t i = 0; i < argCount; ++i)
+        {
+          const CommandLineArgument& lhs_arg = args[i];
+          for(count_t j = 0; j < argCount; ++j)
+          {
+            if(i == j)
+              continue;
+
+            const CommandLineArgument& rhs_arg = args[j];
+            if(rsl::strincmp(lhs_arg.name.data(), rhs_arg.name.data(), lhs_arg.name.length()) == 0)
+            {
+              REX_ERROR(LogEngine, "This executable already has an argument for {} specified in 'g_command_line_args', please resolve the ambiguity", lhs_arg.name);
+              return false;
+            }
+          }
+        }
+
+        return true;
+      }
+
+    private:
+      rsl::vector<ActiveArgument> m_arguments;
```

Comment:
It's it going to work better if we use a hash?
We will look for command line arguments more often than inserting them.

Reply:
yes, but we'll have to hash it every time we want to look it up with kinda defeats the purpose don't you think?

Reply:
what we could do, once compile time sid are in, is pass in an sid and check that.
that'll be hashed at compile time so no runtime overhead